File: 1443_minimumTimeToCollectAllApplesInATree/main.cpp
#include <vector>

class Solution {
private:
    int dfs(int curNode, int parent, std::vector<std::vector<int>>& adjList, std::vector<bool>& hasApple) {
        int time = 0;
        // int childTime;
        for(int child: adjList[curNode]) {
            if(child == parent)
                continue;
            int childTime = dfs(child, curNode, adjList, hasApple);
            if(childTime != 0 || hasApple[child])
                time = time + childTime + 2;
        }
        return time;
    }

public:
    int minTime(int n, std::vector<std::vector<int>>& edges, std::vector<bool>& hasApple) {
        std::vector<std::vector<int>> adjList(n, std::vector<int>());
        for(std::vector<int> edge: edges) {
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
        }
        return dfs(0, -1, adjList, hasApple);
    }
};

File: 1443_minimumTimeToCollectAllApplesInATree/main.py
from typing import List

class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        adj = { i: [] for i in range(n)}
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        def dfs(curNode, parent):
            time = 0

            for child in adj[curNode]:
                if child == parent:
                    continue
                childTime = dfs(child, curNode)
                if childTime or hasApple[child]:
                    time += 2 + childTime
            return time
        
        return dfs(0, -1)
    
Solution().minTime(7, [[0,1],[0,2],[1,4],[1,5],[2,3],[2,6]]
, [False,False,True,False,True,True,False])